File: leetcode/2661_first_completely_painted_row_or_column.py
class Solution:
    def firstCompleteIndex(self, arr: List[int], mat: List[List[int]]) -> int:
        position_map = {}
        rows = len(mat)
        cols = len(mat[0])
        row_count = [cols] * cols
        col_count = [rows] * rows

        # store each num in arr in the position map where it appears in the matrix
        for row in range(rows):
            for col in range(cols):
                position_map[mat[row][col]] = (row, col)

        # loop through the array
        for i in range(len(arr)):
            num = arr[i]
            # get the row and col of the current num in the position map
            row, col = position_map[num]

            # decrement the row count
            row_count[row] -= 1
            # decrement the col count
            col_count[col] -= 1

            # Once either the current row or column count reaches 0, we know it's complete
            if row_count[row] == 0 or col_count[col] == 0:
                return i

        return -1

File: leetcode/2661_first_completely_painted_row_or_column.js
/**
 * @param {number[]} arr
 * @param {number[][]} mat
 * @return {number}
 */
var firstCompleteIndex = function (arr, mat) {
  const rows = mat.length;
  const cols = mat[0].length;
  const positionMap = new Map();
  const rowCount = new Array(rows).fill(cols);
  const colCount = new Array(cols).fill(rows);

  // set our position map so that we can reference it when we loop through the array
  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      positionMap.set(mat[row][col], [row, col]);
    }
  }

  // loop through our array
  for (let i = 0; i < arr.length; i++) {
    const num = arr[i];
    // obtain our row and col coordinate from the position map for the current number
    const [row, col] = positionMap.get(num);
    // decrement the row and col count for the current row and col since we visited it
    rowCount[row]--;
    colCount[col]--;

    // once either the current row or col is 0 then we know we have filled it
    if (rowCount[row] == 0 || colCount[col] == 0) {
      return i;
    }
  }

  return -1;
};

// Time: O(N) where N is the number of elements in the array
// Space: O(N) where N is the number of elements in the array since we store the position of each element in a map